src/Features/LanguageServer/ProtocolUnitTests/CodeActions/CodeActionResolveTests.cs

```diff
@@ -0,0 +1,155 @@
+                kind: CodeActionKind.Refactor,
+                children: Array.Empty<LSP.VSCodeAction>(),
+                data: CreateCodeActionResolveData(CSharpAnalyzersResources.Use_implicit_type, locations["caret"].Single()),
+                diagnostics: null,
+                edit: new LSP.WorkspaceEdit
+                {
+                    DocumentChanges = new TextDocumentEdit[]
+                    {
+                        new TextDocumentEdit
+                        {
+                            TextDocument = new VersionedTextDocumentIdentifier
+                            {
+                                Uri = locations["caret"].Single().Uri
+                            },
+                            Edits = new TextEdit[] {
+                                new TextEdit
+                                {
+                                    NewText = expectedMarkup,
+                                    Range = new LSP.Range { Start = new Position(0, 0), End = new Position(6, 1) }
+                                }
+                            }
+                        }
+                    }
+                });
+
+            var result = await RunGetCodeActionResolveAsync(workspace.CurrentSolution, unresolvedCodeAction);
+            AssertJsonEquals(expected, result);
+        }
+
+        [Fact]
+        public async Task TestCodeActionResolveHandlerAsync_NestedAction()
+        {
+            var initialMarkup =
+@"class A
+{
+    void M()
+    {
+        int {|caret:|}i = 1;
+    }
+}";
+            using var workspace = CreateTestWorkspace(initialMarkup, out var locations);
+
+            var unresolvedCodeAction = CodeActionsTests.CreateCodeAction(
+                title: string.Format(FeaturesResources.Introduce_constant_for_0, "1"),
+                kind: CodeActionKind.Refactor,
+                children: Array.Empty<LSP.VSCodeAction>(),
+                data: CreateCodeActionResolveData(
+                    FeaturesResources.Introduce_constant + string.Format(FeaturesResources.Introduce_constant_for_0, "1"),
+                    locations["caret"].Single()),
+                diagnostics: null);
+
+            var expectedMarkup =
+@"class A
+{
+    private const int V = 1;
+
+    void M()
+    {
+        int i = V;
+    }
+}";
+
+            var expected = CodeActionsTests.CreateCodeAction(
+                title: string.Format(FeaturesResources.Introduce_constant_for_0, "1"),
+                kind: CodeActionKind.Refactor,
+                children: Array.Empty<LSP.VSCodeAction>(),
+                data: CreateCodeActionResolveData(
+                    FeaturesResources.Introduce_constant + string.Format(FeaturesResources.Introduce_constant_for_0, "1"),
+                    locations["caret"].Single()),
+                diagnostics: null,
+                edit: new LSP.WorkspaceEdit
+                {
+                    DocumentChanges = new TextDocumentEdit[]
```

Comment:
let's add a helper to create the workspace edit from markup, uri, and location since this is repeated